src/Avalonia.Controls/Utils/BorderRenderHelper.cs

```diff
@@ -63,17 +70,25 @@ void Update(Size finalSize, Thickness borderThickness, CornerRadius cornerRadius
 
                 if (boundRect.Width != 0 && boundRect.Height != 0)
                 {
-                    var borderGeometryKeypoints =
-                        new BorderGeometryKeypoints(boundRect, borderThickness, cornerRadius, false);
                     var borderGeometry = new StreamGeometry();
```

Comment:
@maxkatz6 I just noticed with #14360 we do actually have geometry Boolean operations implemented. The heavy lifting is done with Skia and it was added in #6576.

I need to use this to calculate the border geometry below. That is what WinUI does and it should make the algorithm exactly the same here. Below (even with the old code) we are relying on an undefined implementation detail to combine the two geometries in the way we want. It's going to be much cleaner not to do this and it should also remove the last little artifacts visible in the exaggerated cases.

![image](https://github.com/AvaloniaUI/Avalonia/assets/17993847/de24ce69-9be0-461d-85e6-687fbdb0f8de)

Reply:
Updated and it looks better now